ZiLing/Midterm-Review.py

```diff
@@ -0,0 +1,481 @@
+Remember that there are two types of numbers 
+int: Intergers, whole numbers 
+float: Decimal numbers 
+
+if you want more math, you have to use the math library 
+
+import math 
+
+then you can do all your wacky math.cos(),math.sine(),math.sqrt() stuff 
+
+Down there are the example of some 
+"""
+# Module is the remain after division
+# for example, 2/2 is 1, there is no remainder
+# 3/2 can also be see as 2+1/2, the remainder will be 1 
+print(2%2)
+print(3%2)
+
+# be aware of the difference in decimal places
+print(2//3)
+print(2/3)
+
+# you can also use + and * on string 
+word = "hello" + " world"
+hohoho = 'ho'*3
+
+"""
+Assignment Operators (in-place operators)
+You dont have to use them, but just know what they are 
+
+= -> x = 5 
++= -> x += 3 is the same sa x = x + 3 
+-= -> x -= 3 is the same as x = x - 3 
+*= -> x *= 3 is the same sa x = x * 3 
+/= -> x /= 3 is the same sa x = x / 3 
+%= -> x %= 3 is the same sa x = x % 3 
+**= -> x **= 3 is the same sa x = x ** 3 
+
+there are more than these, but just know the onces above should be enough 
+for you to go through this course
+
+"""
+
+# You cannot use two assignment operators on the same variable at the same line 
+# But you can mix it with other algebretical operators 
+x += 435+493/2302567
+
+"""
+Everything below will return a boolean value(True or False)
+
+Comparison Oprators 
+
+Operator   Name           Example
+==         Equal          x == y, 2 == 1 
+!=       Not Equal       x != 7, y != y
+>      Greater than       3 > 2, 4 > 5
+<       Less than         1 < 2, 5 < 5 
+>=   Greater or Equal to   1 >= 1, 1 >= 4
+<=   less or equal to      x <= 3 , 4 <= 4 
+
+You usually use them with if statement or while statement
+"""
+
+# You can use more than one comparison operators in the same line 
+for i in range (356,9,-3): 
+    if 325 < i > 35:
+        print (i)
+
+"""
+Logical Oprator 
+
+operator    Returns True if          Example
+and     both statements are true     x < 5 and x > 1 
+or      either statement is true     x < 5 or x > 1 
+not     the statement is false       not(x > 5)
+
+Identity Operator 
+
+operator    Returns True if          Example
+is      both variables are the same   5 is 5, x is 5
+is not  variables are not the same    x is not 4 
+
+Membership Operator 
+
+operator    Returns True if                 Example
+in       object is element in smt        i in mydict
+not in      object not in smth           i not in mydict     
+
+All the operators above return a boolean type value (True or False)
+"""
+
+# You can mix mash all the operators together as long as it makes sense
+for i in range (1,101):
+    if not(x%2 == 0 and x > 5 ): 
+        print (i)
+
+"""
+Conditional statements 
+
+Statement      name   
+if              if      
+elif         else if
+else           else
+while          while 
+"""
+
+"""
+Range
+
+range is a type that gives a range of intergers, like how 
+string is a sequence of characters, range is a sequence 
+of integers
+
```

Comment:
I will add that ranges are not the same as lists. Think of ranges as generators whereas lists actually store values in memory. Because lists in python are implemented at the C-level (Python is built on top of C) as weird array of pointers, I'll try explaining this using C vectors, which you actually have access to in the form of numpy ndarrays.

An array such as `arr = np.array([1,2,3])` creates an object which stores the integers 1, 2, and 3, and their indices in the array. A python list is a weird array of pointers that would point to integers 1, 2, and 3. I'm not sure if python lists use indices or use the C list method of only storing what element comes before or after the current element. A range on the other hand creates an iterable of immutable objects which are not generated and stored in memory. So `rng = range(1,4)` doesn't actually store the values 1, 2, and 3. Instead, these integers are generated when the object (a pointer) is called from the range. Note that ranges can be used to initialize a list, `list(range(1,4))`, and custom numpy arrays require you to use a list to initialize the array. I.e. the list in the following line is first being generated and then the list is used to initialize the array. 
```python
arr = np.array([1,2,3])
```
This is something to keep in mind during optimization. While numpy arrays are generally faster for random access searches and are **very fast** when using numpy's linear algebra operation methods (coded in C and are sometimes multi-threaded for parallel processing which cannot be done in pure python because of the [dreaded] global interpreter lock [GIL]), if you need to perform some simple operations such as iterate through a short list, then even if the iteration is faster with a numpy array, the time taken to initialize the array with a list your create is longer than it would take to simply iterate through the list.